programs/survival/compass.sc

```diff
@@ -0,0 +1,613 @@
+        'item' -> {'type' -> 'item'},
+        'amt' -> {'type' -> 'int', 'min' -> 1, 'max' -> 64, 'suggest' -> [1, 16, 64]},
+        'time' -> {'type' -> 'float', 'min' -> 0.05, 'suggest' -> [0.5, 1, 10]},
+        'enabled' -> {'type' -> 'bool'},
+        'effect' -> {'type' -> 'effect'},
+        'duration' -> {'type' -> 'int', 'min' -> 0, 'max' -> 1000000, 'suggest' -> [15, 60, 1000000]},
+        'amplifier' -> {'type' -> 'int', 'min' -> 0, 'max' -> 255, 'suggest' -> [0, 1, 2, 3, 255]},
+        'dist' -> {'type' -> 'int', 'min' -> 16, 'max' -> 24000000, 'suggest' -> [8, 128, 7000, 24000000]},
+    }
+};
+
+toKeep = parse_nbt(read_file('toKeep', 'nbt'));
+
+if (toKeep == 'null',
+    toKeep = l();
+);
+
+global_toKeep = map(toKeep, l(_: 0, number(_: 1), if(_: 2 == '', null, nbt(_: 2)), _: 3, number(_: 4)));
+
+toKeepEffect = parse_nbt(read_file('toKeepEffect', 'nbt'));
+
+if (toKeepEffect == 'null',
+    toKeepEffect = l();
+);
+
+global_toKeepEffect = map(toKeepEffect, l(_: 0, number(_: 1), number(_: 2), _: 3));
+
+global_logOnTrack = parse_nbt(read_file('logOnTrack', 'nbt'));
+
+if (global_logOnTrack == 'null',
+    global_logOnTrack = false;
+);
+
+global_inventoryData = {};
+
+global_timeToTrack = parse_nbt(read_file('timeToTrack', 'nbt'));
+
+if (global_timeToTrack == 'null',
+    global_timeToTrack = 10;
+);
+
+// Crappy fix for https://github.com/gnembon/fabric-carpet/issues/677
+system_variable_set('compass:timeToTrack', global_timeToTrack);
+
+global_maxCompasses = parse_nbt(read_file('maxCompasses', 'nbt'));
+
+if (global_maxCompasses == 'null',
+    global_maxCompasses = 'infinity';
+);
+
+global_noPortalBreaking = parse_nbt(read_file('noPortalBreaking', 'nbt'));
+
+if (global_noPortalBreaking == 'null',
+    global_noPortalBreaking = false;
+);
+
+global_maxNetherTravel = parse_nbt(read_file('maxNetherTravel', 'nbt'));
+
+if (global_maxNetherTravel == 'null',
+    global_maxNetherTravel = 3000000;
+);
+
+__on_player_dies(player) -> (
+    slots = l(); // The list of the items that need to be given back when the player respawns
+
+    // Find the slots of all the compasses
+    slot = inventory_find(player, 'compass');
+    while (slot != null, 41,
+        // Only add the compass to the list if it's a tracking compass
+        if (inventory_get(player, slot): 2: 'Tracking',
+            triple = inventory_get(player, slot);
+            put(slots, null, l(triple: 0, triple: 1, triple: 2, slot));
+            inventory_set(player, slot, 0);
+        );
+        slot = inventory_find(player, 'compass', slot+1);
+    );
+
+    bySlotNum = map(slots, _: 3); // All the inventory slots that are taken
+
+    // Find all the slots where the player has items that they respawn with
+    for (global_toKeep,
+        if (_: 3 == 'âny' || _: 3 == lower(player ~ 'name'), // It's impossible to use the character ̂  in a player name
```

Comment:
Can't u use null instead?

Reply:
Yeah I could, I just didn't think of it at the time

Reply:
Actually, I was testing it, and I got the error 'cannot reliably encode to a tag the value of 'null'' when saving the spawnWith data, so types are pretty important in this case

Reply:
oh yeah, forgot...

Reply:
wait, try now with the new version? cos if not thats probably a bug